src/test/java/org/apache/cassandra/distributed/api/LogActionTest.java

```diff
@@ -0,0 +1,252 @@
+    }
+
+    @Test
+    public void grepForErrorsNoStacktrace()
+    {
+        LogAction logs = Mockito.mock(LogAction.class, Mockito.CALLS_REAL_METHODS);
+        Mockito.when(logs.match(Mockito.anyLong(), Mockito.any()))
+                .thenAnswer(invoke -> lineIterator(invoke.getArgument(1),
+                        "INFO a",
+                        "INFO b",
+                        "ERROR match",
+                        "INFO d"));
+
+        List<String> matches = logs.grepForErrors().getResult();
+        Assertions.assertThat(matches).isEqualTo(Arrays.asList("ERROR match"));
+    }
+
+    @Test
+    public void grepForErrorsWithStacktrace()
+    {
+        LogAction logs = Mockito.mock(LogAction.class, Mockito.CALLS_REAL_METHODS);
+        Mockito.when(logs.match(Mockito.anyLong(), Mockito.any()))
+                .thenAnswer(invoke -> lineIterator(invoke.getArgument(1),
+                        "INFO a",
+                        "INFO b",
+                        "ERROR match",
+                        "\t\tat class.method(42)",
+                        "\t\tat class.method(42)"));
+
+        List<String> matches = logs.grepForErrors().getResult();
+        Assertions.assertThat(matches).isEqualTo(Arrays.asList("ERROR match\n" +
+                "\t\tat class.method(42)\n" +
+                "\t\tat class.method(42)"));
+    }
+
+    @Test
+    public void grepForErrorsMultilineWarnNotException()
+    {
+        LogAction logs = Mockito.mock(LogAction.class, Mockito.CALLS_REAL_METHODS);
+        Mockito.when(logs.match(Mockito.anyLong(), Mockito.any()))
+                .thenAnswer(invoke -> lineIterator(invoke.getArgument(1),
+                        "INFO a",
+                        "INFO b",
+                        "WARN match",
+                        "\t\tat class.method(42)",
+                        "\t\tat class.method(42)"));
+
+        List<String> matches = logs.grepForErrors().getResult();
+        Assertions.assertThat(matches).isEmpty();
+    }
+
+    @Test
+    public void grepForErrorsWARNWithStacktrace()
+    {
+        LogAction logs = Mockito.mock(LogAction.class, Mockito.CALLS_REAL_METHODS);
+        Mockito.when(logs.match(Mockito.anyLong(), Mockito.any()))
+                .thenAnswer(invoke -> lineIterator(invoke.getArgument(1),
+                        "INFO a",
+                        "INFO b",
+                        "WARN match but exception in test",
+                        "\t\tat class.method(42)",
+                        "\t\tat class.method(42)"));
+
+        List<String> matches = logs.grepForErrors().getResult();
+        Assertions.assertThat(matches).isEqualTo(Arrays.asList("WARN match but exception in test\n" +
+                "\t\tat class.method(42)\n" +
+                "\t\tat class.method(42)"));
+    }
+
+    @Test
+    public void grepForErrorsWARNWithStacktraceFromAssert()
+    {
+        LogAction logs = Mockito.mock(LogAction.class, Mockito.CALLS_REAL_METHODS);
```

Comment:
Looks like most of these lines repeat in many tests here. It may make sense to add a small helper that would receive a multi-line log, then make sure errors match in a specific way.